File: twitter.py
class User:
    def __init__(self, userId, next=None) -> None:
        self.id=userId
        self.posts=None  
        self.following={self}

class Tweet:
    def __init__(self, postId, time, next=None) -> None:
        self.id=postId
        self.timestamp=time
        self.next=None

class Twitter:

    def __init__(self):
        self.time=0
        self.users={}  

    def postTweet(self, userId: int, tweetId: int) -> None:
        self.time+=1
        self.users[userId] = self.users.get(userId, User(userId))
        user=self.users[userId]
        tweet = Tweet(tweetId, self.time)
        tweet.next=user.posts
        user.posts=tweet
        

    def getNewsFeed(self, userId: int) -> List[int]:
        self.time+=1
        self.users[userId] = self.users.get(userId, User(userId))
        user=self.users[userId]
        heap=[] 
        for u in user.following:
            if u.posts:
                heapq.heappush(heap, (-u.posts.timestamp, u.posts))
        i=0
        res=[]
        while i<10 and heap:
            _, post=heapq.heappop(heap)
            res.append(post.id)
            if post.next:
                heapq.heappush(heap, (-post.next.timestamp, post.next))
            i+=1
        return res



    def follow(self, followerId: int, followeeId: int) -> None:
        self.time+=1
        self.users[followerId]=self.users.get(followerId, User(followerId))
        follower= self.users[followerId]
        self.users[followeeId]=self.users.get(followeeId, User(followeeId))
        followee= self.users[followeeId]
        follower.following.add(followee)


    def unfollow(self, followerId: int, followeeId: int) -> None:
        self.time+=1
        self.users[followerId]=self.users.get(followerId, User(followerId))
        follower= self.users[followerId]
        self.users[followeeId]=self.users.get(followeeId, User(followeeId))
        followee= self.users[followeeId]
        follower.following.discard(followee)
        


# Your Twitter object will be instantiated and called as such:
# obj = Twitter()
# obj.postTweet(userId,tweetId)
# param_2 = obj.getNewsFeed(userId)
# obj.follow(followerId,followeeId)
# obj.unfollow(followerId,followeeId)

File: twitter.java
// Time Complexity :O(k log n)
// Space Complexity :O(n)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this :no

public class Twitter {
    private HashMap<Integer, HashSet<Integer>> follows;
    private HashMap<Integer, LinkedList<Tweet>> tweets;
    private int timestamp;
    private class Tweet implements Comparable<Tweet> {
        private int tweetId;
        private int ts;
        private int userId;
        public Tweet(int tweetId, int userId, int timestamp) {
            this.tweetId = tweetId;
            this.ts = timestamp;
            this.userId = userId;
        }
        @Override
        public int compareTo(Tweet t2) {
            return t2.ts - this.ts;
        }
    }
    public Twitter() {
        follows = new HashMap<Integer, HashSet<Integer>>();
        tweets = new HashMap<Integer, LinkedList<Tweet>>();
        timestamp = 0;
    }
    public void postTweet(int userId, int tweetId) {
        if (!tweets.containsKey(userId)) {
            tweets.put(userId, new LinkedList<Tweet>());
        }
        tweets.get(userId).add(0, new Tweet(tweetId, userId, timestamp++));
        if (!follows.containsKey(userId)) {
            follows.put(userId, new HashSet());
            follows.get(userId).add(userId);
        }
    }
     public List<Integer> getNewsFeed(int userId) {
        List<Integer> feed = new LinkedList<Integer>();
        PriorityQueue<Tweet> q = new PriorityQueue<Tweet>();
        if (!follows.containsKey(userId)) {
            return feed;
        }
        HashSet<Integer> followed = follows.get(userId);
        HashMap<Integer, Integer> count = new HashMap<Integer, Integer>();
        for (Integer i:followed) {
            if (tweets.containsKey(i)) {
                Tweet t = tweets.get(i).get(0);
                q.add(t);
                count.put(t.userId, 1);
            }
        }
        while (q.size() > 0 && feed.size() < 10) {
            Tweet t = q.poll();
            feed.add(t.tweetId);
            int next = count.get(t.userId);
            count.put(t.userId, next + 1);
            if (next < tweets.get(t.userId).size()) {
                q.add(tweets.get(t.userId).get(next));
            }
        }
        return feed;
    }
    public void follow(int followerId, int followeeId) {
        if (!follows.containsKey(followerId)) {
            follows.put(followerId, new HashSet());
            follows.get(followerId).add(followerId);
        }
        follows.get(followerId).add(followeeId);
    } public void unfollow(int followerId, int followeeId) {
        if (followerId == followeeId) {
            return;
        }
        if (follows.containsKey(followerId)) {
            follows.get(followerId).remove(followeeId);
        }
    }
}